File: day-19/prg.hs
import System.Environment
import Data.Char
import qualified Data.Map as Map
import Debug.Trace

-- READING MEMORY

type Mem a = Map.Map a a

split1 :: Char -> String -> ([String], String)
split1 _ [] = ([], [])

split1 delim (c:rest)
  | c == delim = ((s:ss), [])
  | space = (ss, s)
  | otherwise = (ss, (c:s))
  where
    (ss, s) = split1 delim rest
    space = isSpace c

split :: Char -> String -> [String]
split c str
  | s == [] = ss
  | otherwise = (s:ss)
  where (ss, s) = split1 c str

readMemory :: (Integral a, Read a) => String -> IO (Mem a)
readMemory fileName = do
  content <- readFile fileName
  let numbers = map read (split ',' content)
  let memory = Map.fromAscList (zip [0..] numbers)
  return memory

-- STATE

data State a = State a a (Mem a) [a] [a] deriving (Show)

state0 :: (Num a) => Mem a -> [a] -> State a
state0 mem input = (State 0 0 mem input [])

decode :: (Integral a ) => a -> a
decode instr = instr `mod` 100

argument :: (Integral a, Show a) => State a -> a -> a
argument (State ip rb mem _ _) i
  | mode == 0 = Map.findWithDefault 0 addr mem
  | mode == 1 = addr
  | mode == 2 = Map.findWithDefault 0 (addr + rb) mem
  | otherwise = error $ "Invalid mode " ++ (show mode)
  where
    mode = ((Map.findWithDefault 0 ip mem) `div` (10 ^ (i + 1))) `mod` 10
    addr = Map.findWithDefault 0 (ip + i) mem

store :: (Integral a, Show a) => State a -> a -> a -> Map.Map a a
store (State ip rb mem _ _) i value
  | mode == 0 = Map.insert addr value mem
  | mode == 2 = Map.insert (addr + rb) value mem
  | otherwise = error $ "Invalid mode " ++ (show mode)
  where
    mode = ((Map.findWithDefault 0 ip mem) `div` (10 ^ (i + 1))) `mod` 10
    addr = Map.findWithDefault 0 (ip + i) mem

add :: (Integral a, Show a) => State a -> State a
add s@(State ip rb _ input output) =
  State (ip + 4) rb mem2 input output
  where
    a1 = argument s 1
    a2 = argument s 2
    mem2 = store s 3 (a1 + a2)

mul :: (Integral a, Show a) => State a -> State a
mul s@(State ip rb _ input output) =
  State (ip + 4) rb mem2 input output
  where
    a1 = argument s 1
    a2 = argument s 2
    mem2 = store s 3 (a1 * a2)

inp :: (Integral a, Show a) => State a -> State a
inp s@(State ip rb _ (i:input) output) =
  State (ip + 2) rb mem2 input output
  where
    mem2 = store s 1 i
inp (State _ _ _ [] _) = error "No input"

outp :: (Integral a, Show a) => State a -> State a
outp s@(State ip rb mem input output) =
  State (ip + 2) rb mem input (o:output)
  where
    o = argument s 1

jmpt :: (Integral a, Show a) => State a -> State a
jmpt s@(State ip rb mem input output) =
  State ip2 rb mem input output
  where
    ip2
      | a1 /= 0 = a2
      | otherwise = ip + 3
    a1 = argument s 1
    a2 = argument s 2

jmpf :: (Integral a, Show a) => State a -> State a
jmpf s@(State ip rb mem input output) =
  State ip2 rb mem input output
  where
    ip2
      | a1 == 0 = a2
      | otherwise = ip + 3
    a1 = argument s 1
    a2 = argument s 2

lt :: (Integral a, Show a) => State a -> State a
lt s@(State ip rb _ input output) =
  State (ip + 4) rb mem2 input output
  where
    a1 = argument s 1
    a2 = argument s 2
    a
      | a1 < a2 = 1
      | otherwise = 0
    mem2 = store s 3 a

eq :: (Integral a, Show a) => State a -> State a
eq s@(State ip rb _ input output) =
  State (ip + 4) rb mem2 input output
  where
    a1 = argument s 1
    a2 = argument s 2
    a
      | a1 == a2 = 1
      | otherwise = 0
    mem2 = store s 3 a

arb :: (Integral a, Show a) => State a -> State a
arb s@(State ip rb mem input output) =
  State (ip + 2) (rb + a1) mem input output
  where
    a1 = argument s 1

halt :: (Integral a, Show a) => State a -> State a
halt (State _ rb mem input output) =
  State (-1) rb mem input output

-- PROCESSING

step :: (Integral a, Show a) => State a -> State a
step s@(State ip _ mem _ _)
  | op == 1 = add s
  | op == 2 = mul s
  | op == 3 = inp s
  | op == 4 = outp s
  | op == 5 = jmpt s
  | op == 6 = jmpf s
  | op == 7 = lt s
  | op == 8 = eq s
  | op == 9 = arb s
  | op == 99 = halt s
  | otherwise = error $ "Invalid instruction " ++ (show op)
  where
    op = decode (Map.findWithDefault 0 ip mem)

halted :: (Num a, Ord a) => State a -> Bool
halted (State ip _ _ _ _) = ip < 0

loopWhile :: (Integral a, Show a) => (State a -> Bool) -> State a -> State a
loopWhile cond s
  | not (halted s) && cond s = loopWhile cond (step s)
  | otherwise = s

-- TASK

affected :: (Integral a, Show a) => Mem a -> (a, a) -> a
affected mem (x, y) = output
  where
    st0 = state0 mem [x, y]
    State _ _ _ _ (output:_) = loopWhile (\_ -> True) st0

task1 :: (Integral a, Show a) => Mem a -> a -> a
task1 mem size = foldl (+) 0 points
  where
    lattice = [(x, y) | x <- [0..size], y <- [0..size]]
    points = map (\xy -> affected mem xy) lattice

nextPosition ::  (Integral a, Show a) => Mem a -> (a, a) -> (a, a)
nextPosition mem (x, y)
  | affected mem (x, y + 1) > 0 = (x, y + 1)
  | affected mem (x + 1, y + 1) > 0 = (x + 1, y + 1)
  | otherwise = (x + 1, y)

searchPosition :: (Integral a, Show a) => Mem a -> a -> (a, a) -> (a, a)
searchPosition mem size (x, y)
  | tr == True = (x, y)
  | otherwise = -- trace ("Position is " ++ (show x) ++ " " ++ (show y))
    (searchPosition mem size (nextPosition mem (x, y)))
  where
    tr = affected mem ((x + size - 1), (y - size + 1)) > 0

task2 :: (Integral a, Show a) => Mem a -> a -> a
task2 mem size = x * 10000 + (y - size + 1)
  where
    (x, y) = searchPosition mem size (0, size - 1)

main :: IO ()
main = do
    (a:_) <- getArgs
    mem <- readMemory a :: IO (Mem Int)
    let count = task1 mem 49
    putStrLn (show count)
    let position = task2 mem 100
    putStrLn (show position)

File: day-21/prg.ml
open Printf

module Mem = Map.Make(struct type t = int let compare = compare end)

type state = State of int * int * int Mem.t * int list * int list

let read_memory filename =
    let chan = open_in filename in
    let content = input_line chan in
    let list = List.map int_of_string (String.split_on_char ',' content) in
    let ilist = List.mapi (fun i a -> (i, a)) list in
    List.fold_left (fun acc (i, m) -> Mem.add i m acc) Mem.empty ilist

let state0 memory input = State (0, 0, memory, input, [])

let step (State (ip, rb, mem, input, output)) =
    let default opt =
        match opt with
        | Some v -> v
        | None -> 0
    in
    let mode_div i =
        match i with
        | 1 -> 100
        | 2 -> 1000
        | 3 -> 10000
        | _ -> raise (Failure "Invalid position")
    in
    let arg (ip, rb, mem) i =
        let op = Mem.find ip mem in
        let mode = (op / (mode_div i) mod 10) in
        let a = Mem.find (ip + i) mem in
        match mode with
        | 0 -> default (Mem.find_opt a mem)
        | 1 -> a
        | 2 -> default (Mem.find_opt (a + rb) mem)
        | _ -> raise (Failure "Invalid arg mode")
    in
    let sto (ip, rb, mem) i value =
        let op = Mem.find ip mem in
        let mode = (op / (mode_div i) mod 10) in
        let a = Mem.find (ip + i) mem in
        match mode with
        | 0 -> Mem.add a value mem
        | 2 -> Mem.add (a + rb) value mem
        | _ -> raise (Failure "Invalid sto mode")
    in
    let add (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State (ip + 4, rb, (sto (ip, rb, mem) 3 (a1 + a2)), input, output)
    in
    let mul (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State (ip + 4, rb, (sto (ip, rb, mem) 3 (a1 * a2)), input, output)
    in
    let inp (State (ip, rb, mem, input, output)) =
        match input with
        | [] -> raise (Failure "No input")
        | (i :: rest) -> State (ip + 2, rb, (sto (ip, rb, mem) 1 (i)), rest, output)
    in
    let outp (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        printf "%c" (if a1 < 255 then char_of_int a1 else '%');
        State (ip + 2, rb, mem, input, a1 :: output)
    in
    let jmpt (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State ((if a1 <> 0 then a2 else ip + 3), rb, mem, input, output)
    in
    let jmpf (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State ((if a1 = 0 then a2 else ip + 3), rb, mem, input, output)
    in
    let lt (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State (ip + 4, rb, (sto (ip, rb, mem) 3 (if a1 < a2 then 1 else 0)), input, output)
    in
    let eq (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        let a2 = arg (ip, rb, mem) 2 in
        State (ip + 4, rb, (sto (ip, rb, mem) 3 (if a1 = a2 then 1 else 0)), input, output)
    in
    let arb (State (ip, rb, mem, input, output)) =
        let a1 = arg (ip, rb, mem) 1 in
        State (ip + 2, rb + a1, mem, input, output)
    in
    let halt (State (_, rb, mem, input, output)) =
        State (-1, rb, mem, input, output)
    in

    let op = (Mem.find ip mem) mod 100 in
    match op with
    | 1 -> add (State (ip, rb, mem, input, output))
    | 2 -> mul (State (ip, rb, mem, input, output))
    | 3 -> inp (State (ip, rb, mem, input, output))
    | 4 -> outp (State (ip, rb, mem, input, output))
    | 5 -> jmpt (State (ip, rb, mem, input, output))
    | 6 -> jmpf (State (ip, rb, mem, input, output))
    | 7 -> lt (State (ip, rb, mem, input, output))
    | 8 -> eq (State (ip, rb, mem, input, output))
    | 9 -> arb (State (ip, rb, mem, input, output))
    | 99 -> halt (State (ip, rb, mem, input, output))
    | _ -> raise (Failure "Invalid instruction")

let halted (State (ip, _, _, _, _)) = ip < 0

let rec loop_while cond state =
    if not (halted state) && cond state then
        loop_while cond (step state)
    else
        state

let droid memory program =
    let chars = List.init (String.length program) (fun i -> int_of_char (String.get program i)) in
    let (State (ip, _, _, _, output)) = loop_while (fun st -> true) (state0 memory chars) in
    let (o :: _) = output in
    printf "%d\n" o

let () =
    let filename = Sys.argv.(1) in
    let memory = read_memory filename in
    let t1 = "NOT A T\nOR T J\nNOT B T\nOR T J\nNOT C T\nOR T J\nAND D J\nWALK\n" in
    droid memory t1;
    let t2 = "OR E J\nAND F J\nAND G J\nNOT H T\nOR T J\nOR C J\nOR E T\nAND I T\nOR B T\nAND T J\nAND A J\nNOT J J\nAND D J\nRUN\n" in
    droid memory t2
    (* Mem.iter (fun a b -> printf "%d %d\n" a b) memory; *)

